### PR TITLE
Fix constantface comm

### LINKED ISSUE #541: Constant Face Boundary Conditions do not communicate properly
The CG solver in IPPL based on finite differences does not communicate halo cells properly. The relative error stays constant across all grid sizes.

I created a new test file for each test run here, I'll add them in the PR. Each test was run with 4 CPUs. Single Rank the results are as expected and converge to the correct solution.

TestCGSolver_convergence_constant:

``` csv
size,h,relError,trueResidual,solverResidual,itCount,solveTime
4,0.2,0.5140887128517535,3.820123950655526e-16,4.878124942680895e-16,6,0.0009340000000000008
8,0.1111111111111111,0.6288001761049367,3.450766361756485e-15,6.512311282577119e-13,25,0.0002620000000000001
16,0.05882352941176471,0.6976987791335945,9.667134093282123e-14,5.012346698362658e-11,50,0.0008530000000000013
32,0.0303030303030303,0.735402974261647,8.992190749713994e-14,1.160814327987606e-10,101,0.006244999999999999
64,0.01538461538461539,0.7550241557554113,2.101455100276474e-13,3.71812280368397e-10,205,0.07481400000000001
128,0.007751937984496124,0.7650073441869345,1.029147810596784e-12,1.046144106420834e-09,414,1.095381
```
Image

TestCGSolver_convergence_periodic:
``` csv
size,h,relError,trueResidual,solverResidual,itCount,solveTime
4,0.25,0.6377366121760882,1.596057071843241e-16,5.123796534383003e-14,1,0.0008059999999999994
8,0.125,0.06610999025353599,4.70268102220917e-16,5.55931194985223e-14,4,8.000000000000021e-05
16,0.0625,0.01511692691447543,5.860301855464142e-11,1.112753509887523e-07,12,0.0003029999999999995
32,0.03125,0.003724276265085818,5.170018915267303e-11,2.776629330922694e-07,21,0.001708999999999999
64,0.015625,0.0009277485090611948,9.456697119791474e-11,1.436507970115203e-06,41,0.016218
128,0.0078125,0.0002317313213305733,8.187506446496449e-11,3.51770745253475e-06,81,0.225149
```
Image

The first step is worse, as the resolution of the grid is to coarse to properly show the SinSin waves.

### PR DESCRIPTION
**Pull request by NickvonWit**
### Fix face-specific boundary detection in `ExtrapolateFace::apply()`

This PR fixes #541. It also adds problem specific, Constant or Periodic Face, test files for the CG Solver.

The new test file for the constant face accounts for #535 by shifting the coordinates, to have the halo cells lie on the boundary instead of. The mechanism is explained in detail in my thesis.

**File:** `src/Field/BcTypes.hpp` (`ExtrapolateFace<Field>::apply()`)

The old boundary check used an OR over both faces, so a rank touching the lower
global boundary would also apply the upper face BC (and vice versa). On multiple
ranks this overwrote internal halo cells that should come from neighbouring
ranks, breaking `ConstantFace` / `ZeroFace` problems.

The check is now face-specific: a face is only treated as a physical boundary if
this rank owns the matching global boundary side. Internal interfaces fall
through to halo exchange as intended.

**Before:**
```cpp
bool isBoundary = (lDomains[myRank][d].max() == domain[d].max())
                  || (lDomains[myRank][d].min() == domain[d].min());
```
**After:**
```cpp
const bool isLowerFace = !(face & 1);

const bool isBoundary =
    isLowerFace ? (lDomains[myRank][d].min() == domain[d].min())
                : (lDomains[myRank][d].max() == domain[d].max());
```

Running the same tests as in #541, again with 4 ranks/CPUs:

TestCGSolver_convergence_constant:
```csv
size,h,relError,trueResidual,solverResidual,itCount,solveTime
4,0.2,0.03993833615825652,5.192396279594881e-16,1.548388472317241e-14,4,0.0007419999999999927
8,0.1111111111111111,0.0118946827628043,2.280480765191515e-14,4.454411477477602e-12,18,0.0001960000000000017
16,0.05882352941176471,0.003297211400358864,7.361141393128399e-14,3.744523593663216e-11,30,0.001140999999999989
32,0.0303030303030303,0.00087231155428423,1.311740466831537e-13,1.40521600427478e-10,57,0.008532999999999999
64,0.01538461538461539,0.000224653947278718,4.601852170306531e-13,3.315314961627911e-10,113,0.041592
128,0.007751937984496124,5.702542539421606e-05,2.503480591224792e-12,1.025120149851657e-09,226,0.622335
```
cg_fix_const

TestCGSolver_convergence_periodic:
```csv
size,h,relError,trueResidual,solverResidual,itCount,solveTime
4,0.25,0.6377366121760882,1.596057071843241e-16,5.123796534383003e-14,1,0.0009799999999999948
8,0.125,0.06610999025353599,4.70268102220917e-16,5.55931194985223e-14,4,8.000000000001062e-05
16,0.0625,0.01511692691447543,5.860301855464142e-11,1.112753509887523e-07,12,0.0002899999999999986
32,0.03125,0.003724276265085818,5.170018915267303e-11,2.776629330922694e-07,21,0.001666000000000001
64,0.015625,0.0009277485090611948,9.456697119791474e-11,1.436507970115203e-06,41,0.01862800000000001
128,0.0078125,0.0002317313213305733,8.187506446496449e-11,3.51770745253475e-06,81,0.232264
```
cg_fix_per

### CODE REVIEW

File: test/solver/TestCGSolver_convergence_constant.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        int maxPow        = 7;
+        double tolerance  = 1e-13;
+        int maxIterations = 4000;
+
+        if (argc > 1 && argv[1][0] != '-') {
+            maxPow = std::atoi(argv[1]);
+        }
+
+        Inform m("");
+        m << "size,h,relError,trueResidual,solverResidual,itCount,solveTime" << endl;
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_constant.cpp]: I think here it's better to print only the same things as were being printed by the TestCGSolver_convergence before.

File: test/solver/TestCGSolver_convergence_constant.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+                    const double cax = cos(ax);
+                    const double cay = cos(ay);
+                    const double caz = cos(az);
+
+                    const double u = sx * sy * sz;
+
+                    const double f = pi * pi
+                                     * (cx * sax * sy * sz + sx * cy * say * sz + sx * sy * cz * saz
+                                        + sx * sy * sz * (cax * cax + cay * cay + caz * caz));
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_constant.cpp]: maybe change it back to the same function as the original file, for readability

File: test/solver/TestCGSolver_convergence_constant.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+            error                     = -laplace(lhs) - rhs;
+            const double trueResidual = norm(error) / norm(rhs);
+
+            const double solverResidual = solver.getResidue();
+            const int itCount           = solver.getIterationCount();
+            const double solveTime      = t1 - t0;
+
+            m << pt << "," << std::setprecision(16) << dx << "," << relError << "," << trueResidual
+              << "," << solverResidual << "," << itCount << "," << solveTime << endl;
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_constant.cpp]: as said before, change back to original output results

File: test/solver/TestCGSolver_convergence_periodic.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        int maxPow        = 7;
+        double tolerance  = 1e-10;
+        int maxIterations = 4000;
+
+        if (argc > 1 && argv[1][0] != '-') {
+            maxPow = std::atoi(argv[1]);
+        }
+
+        Inform m("");
+        m << "size,h,relError,trueResidual,solverResidual,itCount,solveTime" << endl;
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_periodic.cpp]: same comment as before regarding which results are printed

File: test/solver/TestCGSolver_convergence_periodic.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+                    const double cax = cos(ax);
+                    const double cay = cos(ay);
+                    const double caz = cos(az);
+
+                    const double u = sx * sy * sz;
+
+                    const double f = a * a
+                                     * (cx * sax * sy * sz + sx * cy * say * sz + sx * sy * cz * saz
+                                        + sx * sy * sz * (cax * cax + cay * cay + caz * caz));
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_periodic.cpp]: same comments as above

File: test/solver/TestCGSolver_convergence_periodic.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+            error                     = -laplace(lhs) - rhs;
+            const double trueResidual = norm(error) / norm(rhs);
+
+            const double solverResidual = solver.getResidue();
+            const int itCount           = solver.getIterationCount();
+            const double solveTime      = t1 - t0;
+
+            m << pt << "," << std::setprecision(16) << dx << "," << relError << "," << trueResidual
+              << "," << solverResidual << "," << itCount << "," << solveTime << endl;
```
Comment (s-mayani) [inline test/solver/TestCGSolver_convergence_periodic.cpp]: same comment as above

**Review by s-mayani — CHANGES_REQUESTED** [review]

**Comment by NickvonWit** [comment]
kept the h in the output as it changes per test, to achieve second order convergence in the constant face test.

**Review by s-mayani — APPROVED** [review]

**Comment by s-mayani** [comment]
cscs-ci run cscs-ci-mi300, cscs-ci-gh200, cscs-ci-openmp

**Comment by aaadelmann** [comment]
cscs-ci run cscs-ci-mi300, cscs-ci-gh200, cscs-ci-openmp

**Comment by aaadelmann** [comment]
cscs-ci run cscs-ci-mi300, cscs-ci-gh200, cscs-ci-openmp

**Comment by aaadelmann** [comment]
cscs-ci run cscs-ci-mi300 cscs-ci-gh200 cscs-ci-openmp

**Comment by biddisco** [comment]
cscs-ci run cscs-ci-mi300, cscs-ci-gh200, cscs-ci-openmp
